src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <Columns/ColumnArray.h>
+#include <DataTypes/DataTypeArray.h>
+
+#include <AggregateFunctions/AggregateFunctionGroupSortedArrayData.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+
+namespace DB
+{
+template <typename T, bool is_plain_column>
+inline T readItem(const IColumn * column, Arena * arena, size_t row)
+{
+    if constexpr (std::is_same_v<T, StringRef>)
+    {
+        if constexpr (is_plain_column)
+        {
+            StringRef str = column->getDataAt(row);
+            auto ptr = arena->alloc(str.size);
+            std::copy(str.data, str.data + str.size, ptr);
+            return StringRef(ptr, str.size);
+        }
+        else
+        {
+            const char * begin = nullptr;
+            return column->serializeValueIntoArena(row, *arena, begin);
+        }
+    }
+    else
+    {
+        return column->getUInt(row);
+    }
+}
+
+template <typename T>
+void getFirstNElements(const T * data, int num_elements, int threshold, size_t * results)
+{
+    for (int i = 0; i < threshold; i++)
+    {
+        results[i] = 0;
+    }
+
+    threshold = std::min(num_elements, threshold);
+    int current_max = 0;
+    int cur;
+    int z;
+    for (int i = 0; i < num_elements; i++)
+    {
+        //Starting from the highest values and we look for the immediately lower than the given one
+        for (cur = current_max; cur > 0 && (data[i] < data[results[cur - 1]]); cur--)
+            ;
+
+        if (cur < threshold)
+        {
+            //Move all the higher values 1 position to the right
+            for (z = current_max - 1; z >= cur; z--)
+                results[z] = results[z - 1];
```

Comment:
Also, this seems to cause heap-buffer-overflow in [some](https://s3.amazonaws.com/clickhouse-test-reports/34055/451d0c09bf158bf2e6a4239781ca2ae261eb192d/fuzzer_astfuzzerasan,actions//report.html) cases.

Reply:
Thanks, I am fixing it.